isb_web/isb_solr_query.py

```diff
@@ -0,0 +1,222 @@
+import typing
+
+import requests
+import geojson
+
+import isb_web.config
+
+BASE_URL = isb_web.config.Settings().solr_url
+_RPT_FIELD = "producedBy_samplingSite_location_rpt"
+
+# Identify the bounding boxes for solr and leaflet for diagnostic purposes
+SOLR_BOUNDS = -1
+LEAFLET_BOUNDS = -2
+
+MIN_LAT = "min_lat"
+MAX_LAT = "max_lat"
+MIN_LON = "min_lon"
+MAX_LON = "max_lon"
+
+# 0.2 seems ok for the grid cells.
+_GEOJSON_ERR_PCT = 0.2
+# 0.1 for the leaflet heatmap tends to generate more cells for the heatmap “blob” generation
+_LEAFLET_ERR_PCT = 0.1
+
+
+def _get_heatmap(
+    q: typing.AnyStr, bb: typing.Dict, dist_err_pct: float, grid_level=None
+) -> typing.Dict:
+    # TODO: dealing with the antimeridian ("dateline") in the Solr request.
+    # Should probably do this by computing two requests when the request BB
+    # straddles the antimeridian.
+    if bb is None or len(bb) < 2:
+        bb = {MIN_LAT: -180.0, MAX_LAT: 180.0, MIN_LON: -90.0, MAX_LON: 90.0}
+    if bb[MIN_LAT] < -180.0:
+        bb[MIN_LAT] = -180.0
+    if bb[MAX_LAT] > 180.0:
+        bb[MAX_LAT] = 180.0
+    # logging.warning(bb)
+    headers = {"Accept": "application/json"}
+    params = {
+        "q": q,
+        "rows": 0,
+        "wt": "json",
+        "facet": "true",
+        "facet.heatmap": _RPT_FIELD,
+        "facet.heatmap.distErrPct": dist_err_pct,
+        # "facet.heatmap.gridLevel": grid_level,
+        "facet.heatmap.geom": f"[{bb[MIN_LAT]} {bb[MIN_LON]} TO {bb[MAX_LAT]} {bb[MAX_LON]}]",
+    }
+    # if grid level is None, then Solr calculates an "appropriate" grid scale
+    # based on the bounding box and distErrPct. Seems a bit off...
+    if grid_level is not None:
+        params["facet.heatmap.gridLevel"] = grid_level
+    # Get the solr heatmap for the provided bounds
+    url = f"{BASE_URL}/select"
+    response = requests.get(url, headers=headers, params=params)
+
+    # logging.debug("Got: %s", response.url)
+    res = response.json()
+    hm = res.get("facet_counts", {}).get("facet_heatmaps", {}).get(_RPT_FIELD, {})
+    return hm
+
+
+##
+# Create a GeoJSON rendering of the Solr Heatmap response.
+# Generates a GeoJSON polygon (rectangle) feature for each Solr heatmap cell
+# that has a count value over 0.
+# Returns the generated features as a GeoJSON FeatureCollection
+#
+def solr_geojson_heatmap(
+    q, bb, grid_level=None, show_bounds=False, show_solr_bounds=False
+):
+    hm = _get_heatmap(q, bb, _GEOJSON_ERR_PCT, grid_level)
+    # print(hm)
+    gl = hm.get("gridLevel", -1)
+    # logging.warning(hm)
+    d_lat = hm["maxY"] - hm["minY"]
+    dd_lat = d_lat / (hm["rows"])
+    d_lon = hm["maxX"] - hm["minX"]
+    dd_lon = d_lon / (hm["columns"])
+    lat_0 = hm["maxY"]  # - dd_lat
+    lon_0 = hm["minX"]  # + dd_lon
+    _max_value = 0
+
+    # Container for the generated geojson features
+    grid = []
+    if show_bounds:
+        bbox = geojson.Feature(
+            geometry=geojson.Polygon(
```

Comment:
PyCharm is unhappy about the typing here and is flagging these warnings:

```
Expected type 'Union[tuple, list[tuple]]', got 'list[list[tuple[Any, Any]]]' instead
```

@datadavev do you know if these are legit problems?